### PR TITLE
Editable semantic tokens on top of #135

### PR DESCRIPTION
**Pull request by tinkererlife**
# submission name:

editable_semantic_tokens

# upload zipped `archive.zip`

[Download archive.zip](https://github.com/tinkererlife/comma_video_compression_challenge/releases/download/editable-semantic-tokens-p8/archive.zip)

185,653 bytes.

SHA-256:
```text
0e496aea27948d23128fc9292655fc784715c6366d2d0fa582248a3d01f561bd
```

# report.txt

Local T4 validation, pending official evaluation. Full-precision results:

```text
=== Evaluation results over 600 samples ===
  Average PoseNet Distortion: 6.13140400674e-06
  Average SegNet Distortion: 0.000291451375233
  Submission file size: 185,653 bytes
  Original uncompressed size: 37,545,489 bytes
  Compression Rate: 0.00494474848896
  Final score: 100*segnet_dist + √(10*posenet_dist) + 25*rate = 0.160594177969
```

# does your submission require gpu for evaluation (inflation)?

Yes. Please use `linux-nvidia-t4`.

# did you include the compression script? and want it to be merged?

No. This submission includes the decoder and archive checksum, with the compressed archive provided as a release asset.

# is this submission competitive or innovative? explain why

The local T4 score is 0.160594, approximately 1.03% lower than #135's reported 0.162268, pending official confirmation.

The contribution is an encoder-side optimization: treating semantic maps as editable tokens and optimizing their combined perception and compression cost. The decoder architecture and renderer weights remain inherited from #135.

# additional comments

This builds on #135, which builds on #130.

The base idea is to treat the semantic maps as editable instead of requiring them to match SegNet's original output exactly. A token change might make the perception score worse, but is still accepted if the reduction in archive size outweighs that. The modified maps are then encoded losslessly.

## Method

The experiments included proposals based only on HPAC coding costs, as well as gradient-assisted proposals. Later passes combined estimated coding savings with a gradient-based estimate of the perception change.

For the gradient estimate, I backpropagated through the fixed renderer and SegNet/PoseNet models to the token representation. This helped rank possible replacements without rendering every alternative. Combining that estimate with coding costs also allowed candidates that worsened perception but saved enough bits.

The shortlisted discrete changes were then actually rendered and evaluated, with a more detailed calculation of the affected coding costs. Changes were accepted when the combined objective improved.

I ran repeated sweeps over the frames, testing individual token replacements and selected pairs. Pairs allow coordinated changes that may not help individually. 

Between token sweeps, I also adjusted the carrier parameters, which control the synthetic first image used by PoseNet, to better match the modified frames.

I also fine-tuned HPAC's probability model on the edited token maps. 

## Decoder changes

GitHub shows 25 added files because this submission uses a new directory. These are 24 decoder code files and one archive checksum file, not a new decoder implementation.

The included decoder files are unchanged from #135 except for the expected archive hash and size in `inflate.py`. The optimization results are stored in the archive.

The three commits separate:

1. Importing the required decoder files from #135.
2. Updating the archive-specific constants and adding the checksum.
3. Renaming the submission directory.

## AI assistance

Codex/claude provided substantial assistance with research, implementation, experiment execution, debugging, and validation.

### CODE REVIEW

**Comment by YassineYousfi** [comment]
congrats!
